### PR TITLE
texture caching (fixes #606)

### LINKED ISSUE #606: cache textures in material component
<a href="https://github.com/ngokevin"></img></a> **Issue by [ngokevin](https://github.com/ngokevin)**
_Tuesday Nov 03, 2015 at 02:20 GMT_
_Originally opened as https://github.com/aframevr/aframe-core/issues/377_

----

### PR DESCRIPTION
**Pull request by ngokevin**
Image texture cache. Image source -> UV repeat -> texture.

```js
{
  'texture.png': {
    '1 1': THREE.Texture(),
    '2 8': THREE.Texture()
  }
}
```

Video texture cache. Video hash -> texture.

```js
{
  'src:myvideo.mp4;width:500;': THREE.Texture(),
  'myvideoid': THREE.Texture()
}
```

### CODE REVIEW

File: src/utils/texture.js [hunk 1]
```diff
@@ -1,68 +1,206 @@
...
+ */
 var debug = require('./debug');
-var error = debug('components:texture:error');
 var THREE = require('../lib/three');
+
+var EVENTS = {
+  TEXTURE_LOADED: 'material-texture-loaded'
+};
+var error = debug('components:texture:error');
+var textureCache = {};
```
Comment (dmarcos) [inline src/utils/texture.js]: The texture cache should probably part of the material system

Comment (ngokevin) [inline src/utils/texture.js]: Can move code to use systems (along with sound, light, camera code) after release.

**Comment by ngokevin** [comment]
Ready for some perf wins.

**Comment by dmarcos** [comment]
Do we have any example that shows the perf wins with this patch?

**Comment by ngokevin** [comment]
The tests show there are less ```THREE.Texture()```s being created thus less textures being pushed to the GPU. I have an issue on file to add more stats to the RStats.

I'll ask DrawVR to get the Minecraft example in since that reused a lot of textures. And also to test it with the crashing city scene.

File: src/utils/texture.js [hunk 2]
```diff
@@ -90,73 +236,40 @@ function createVideoEl (material, src, width, height) {
...
+  if (videoEl.getAttribute('loop') === 'false') {
+    videoEl.removeAttribute('loop');
+  }
   if (videoEl.getAttribute('preload') === 'false') {
     videoEl.preload = 'none';
   }
   return videoEl;
 }
 
-function loadImage (material, data, src) {
```
Comment (dmarcos) [inline src/utils/texture.js]: Please don't move functions randomly because it makes reviews much harder

**Pull request merged**